port.js

```diff
@@ -169,52 +169,45 @@ Port.prototype.stop = function stop() {
 };
 
 Port.prototype.request = function request() {
-    var port = this;
-    var $meta = (arguments.length && arguments[arguments.length - 1]) || {};
+    var $meta = arguments.length && arguments[arguments.length - 1];
+    if (!arguments.length) {
+        return Promise.reject(errors.missingParams());
+    } else if (!$meta) {
+        return Promise.reject(errors.missingMeta());
+    }
     var args = Array.prototype.slice.call(arguments);
-    return when.promise(function requestPromise(resolve, reject) {
-        if (!args.length) {
-            reject(errors.missingParams());
-        } else if (!$meta) {
-            reject(errors.missingMeta());
-        } else {
-            $meta.callback = function requestPromiseCb(msg) {
-                if ($meta && $meta.mtid !== 'error') {
-                    resolve(Array.prototype.slice.call(arguments));
-                } else {
-                    reject(msg);
-                }
-                return true;
-            };
-            if (this.queue) {
-                this.queue.add(args);
-            } else if ($meta && $meta.conId && this.queues[$meta.conId]) {
-                this.queues[$meta.conId].add(args);
-            } else if (Object.keys(this.queues).length && port.connRouter && typeof port.connRouter === 'function') {
-                var queue = this.queues[port.connRouter(this.queues, Array.prototype.slice.call(arguments))];
-                queue && queue.add(args);
+    var queue = this.queue || this.queues[$meta.conId] || this.queues[this.connRouter(this.queues, args)];
+    if (!queue) {
+        return Promise.reject(errors.notConnected(this.config.id));
+    }
+    queue.add(args);
```

Comment:
the message has to be queued after $meta.callback is assigned. Doing it virse versa might cause unexpected behavior.